### PR TITLE
Check manifest fields for type errors.

### PR DESCRIPTION
**Pull request by dblock**
Signed-off-by: dblock <dblock@dblock.org>

### Description

- Add type annotations and type checking to manifest fields, catching `ref` values that are floats instead of strings.
- Bring in dataclass type annotation checker from https://gist.github.com/svermeulen/8a6ad727113a6d07f78332152c1d33b2, this should really be PRed upstream to https://github.com/agronholm/typeguard/issues/161, I might take it up.
- Inherit `TestManifest` from `Manifest`, enabling `from_file` and `from_path`, add tests.
 
### Issues Resolved

Closes https://github.com/opensearch-project/opensearch-build/issues/470.

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: bundle-workflow/src/manifests/typechecked.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (owaiskazi19) [inline bundle-workflow/src/manifests/typechecked.py:1]: Not sure but can we have the name of the file as `typechecker` as `typechecked` is an internal keyword?

File: bundle-workflow/tests/tests_manifests/data/invalid-ref.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+build:
+  name: OpenSearch
+  version: 1.1.0
+components:
+  - name: OpenSearch
+    repository: https://github.com/opensearch-project/OpenSearch.git
+    ref: 1.1
+    checks:
+      - gradle:publish
+      - gradle:properties:version
```
Comment (owaiskazi19) [inline bundle-workflow/tests/tests_manifests/data/invalid-ref.yml:12]: New line here

**Review by owaiskazi19 — APPROVED** [review]
Few should have changes

**Review by peternied — CHANGES_REQUESTED** [review]
This seems like a large/complex amount of work when all the types in the build manifest are strings. Is there a usecase where we need to support numeric values?

Could we do something simpler where we automatically convert all values to strings?

**Comment by dblock** [comment]
> This seems like a large/complex amount of work when all the types in the build manifest are strings. Is there a usecase where we need to support numeric values?
> 
> Could we do something simpler where we automatically convert all values to strings?

I see the following options.

1. We could not merge this and `str()` the `ref` on a git repo. This would be a hack to work around the fact that 1.1 is of type float and not a string, and there's no guarantee that a branch name such as 1.01 doesn't suddenly convert to a string `1.099999` or something like that in the future, which will be hard to debug.
2. We could check explicitly that this one field is an `str` instead of adding a typechecker. Feels not very future-proof. Plus the check as is also catches issues such as not specifying a dictionary where it should be one - it's an actual validator for manifest schema.

WDYT?

File: bundle-workflow/tests/tests_manifests/data/invalid-ref.yml (hunk 2 above)
Comment (dblock) [reply bundle-workflow/tests/tests_manifests/data/invalid-ref.yml]: Added.

File: bundle-workflow/src/manifests/typechecked.py (hunk 1 above)
Comment (dblock) [reply bundle-workflow/src/manifests/typechecked.py]: Renamed to match the class name, `dataclass_typechecked`.

**Comment by dblock** [comment]
@peternied LMK what you think - this does fix a bug and prevents future bugs.

**Comment by peternied** [comment]
What about defining a schema document and validating?  Implemented something close on my fork [Schematize and validate bundle manitests](https://github.com/peternied/opensearch-build/commit/3eecbdd74343457f02a25c80e5396f02bbc654b9)

**Review by peternied — APPROVED** [review]
Since python doesn't enforce typing I guess I am hesitant to rely on the current effort because while it works right now, a new untyped field could be added and it might be missed.

**Comment by dblock** [comment]
I like the schema idea better. Do you want to finish that?

**Comment by peternied** [comment]
@dblock Could you pick it up the schema change and make a new PR?

**Review by owaiskazi19 — APPROVED** [review]

**Comment by dblock** [comment]
Closing in favor of https://github.com/opensearch-project/opensearch-build/pull/509.

**Pull request closed without merging**